File: runt/apis/cldc-compact/java/util/Random.java
// -*- Mode: Java; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// Multi-Phasic Applications: SquirrelJME
//     Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
//     Copyright (C) Multi-Phasic Applications <multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the GNU General Public License v3+, or later.
// See license.mkd for licensing and copyright information.
// ---------------------------------------------------------------------------

package java.util;

import cc.squirreljme.runtime.cldc.annotation.ImplementationNote;

/**
 * This is an insecure pseudo-random number generator.
 *
 * This class is thread safe however if multiple threads are using a single
 * instance then it is undefined if their order is the same.
 *
 * @since 2018/11/02
 */
@ImplementationNote("This code essentially is a copy and paste from the " +
	"library documentation since it must always be the same exact code. " +
	"Since there is only one way to implement this class it should be " +
	"legal to do so for compatibility purposes.")
public class Random
{
	/** The current seed value. */
	private long _seed;
	
	/** The next gaussian value. */
	private double _nextg;
	
	/** Has next next gaussian value? */
	private boolean _hasnng;
	
	/**
	 * Initializes the random number generator using an unspecified and
	 * potentially different key per construction.
	 *
	 * @since 2018/11/02
	 */
	public Random()
	{
		this(System.currentTimeMillis() +
			System.nanoTime());
	}
	
	/**
	 * Initializes the random number generator with the given seed.
	 *
	 * @param __seed The seed to initialize with.
	 * @since 2018/11/02
	 */
	public Random(long __seed)
	{
		this.setSeed(__seed);
	}
	
	/**
	 * Generates the next pseudorandom number, this method is used by every
	 * other method in this class. As such this is the only method that has
	 * to be changed if the internal algorithm is to be changed.
	 *
	 * @param __bits The number of bits to return.
	 * @since 2018/11/02
	 */
	protected int next(int __bits)
	{
		// This has to be atomic
		synchronized (this)
		{
			// Update the seed
			long seed = (this._seed * 0x5DEECE66DL + 0xBL) & ((1L << 48) - 1);
			this._seed = seed;
			
			return (int)(seed >>> (48 - __bits));
		}
	}
	
	/**
	 * Returns the next boolean value.
	 *
	 * @return The next value.
	 * @since 2018/11/02
	 */
	public boolean nextBoolean()
	{
		return this.next(1) != 0;
	}
	
	/**
	 * Generates random bytes and places them into the array.
	 *
	 * @param __a The output array.
	 * @throws NullPointerException On null arguments.
	 * @since 2018/11/02
	 */
	public void nextBytes(byte[] __a)
		throws NullPointerException
	{
		if (__a == null)
			throw new NullPointerException("NARG");
		
		for (int i = 0, l = __a.length; i < l;)
			for (int r = this.nextInt(), n = Math.min(l - i, 4); n-- > 0;
				r >>= 8)
				__a[i++] = (byte)r;
	}
	
	/**
	 * Returns the next double value.
	 *
	 * @return The next value.
	 * @since 2018/11/02
	 */
	public double nextDouble()
	{
		return (((long)this.next(26) << 27) + this.next(27)) /
			(double)(1L << 53);
	}
	
	/**
	 * Returns the next float value.
	 *
	 * @return The next value.
	 * @since 2018/11/02
	 */
	public float nextFloat()
	{
		return this.next(24) / ((float)(1 << 24));
	}
	
	/**
	 * Return the next gaussian value.
	 *
	 * @return The next value.
	 * @since 2018/11/02
	 */
	public double nextGaussian()
	{
		synchronized (this)
		{
			// If there was already a cached value, use it
			boolean hasnng = this._hasnng;
			if (hasnng)
			{
				this._hasnng = false;
				return this._nextg;
			}
			
			// Otherwise generate it
			else
			{
				double v1, v2, s;
				do
				{
					// Between -1.0 and 1.0
					v1 = 2 * nextDouble() - 1;
					v2 = 2 * nextDouble() - 1;

					s = v1 * v1 + v2 * v2;
				} while (s >= 1 || s == 0);
				
				double multiplier =
					Random.__strictSqrt(-2 * Random.__strictLog(s) / s);
				
				// Store for next time
				this._nextg = v2 * multiplier;
				this._hasnng = true;
				
				// Return generated value
				return v1 * multiplier;
			}
		}
	}
	
	/**
	 * Returns the next int value.
	 *
	 * @return The next value.
	 * @since 2018/11/02
	 */
	public int nextInt()
	{
		return this.next(32);
	}
	
	/**
	 * Returns the next int value between inclusive 0 and exclusive
	 * {@code __cap}.
	 *
	 * @param __cap The exclusive value to end at.
	 * @return The next value.
	 * @throws IllegalArgumentException If the cap is zero or negative.
	 * @since 2018/11/02
	 */
	public int nextInt(int __cap)
		throws IllegalArgumentException
	{
		// {@squirreljme.error ZZ2x Cap value must be a positive integer}
		if (__cap <= 0)
			throw new IllegalArgumentException("ZZ2x");
		
		// Cap is a power of two
		if ((__cap & -__cap) == __cap)
			return (int)((__cap * (long)this.next(31)) >> 31);
		
		// Otherwise calculate based on values in the range
		int bits, val;
		do
		{
			bits = this.next(31);
			val = bits % __cap;
		} while (bits - val + (__cap - 1) < 0);
		
		return val;
	}
	
	/**
	 * Returns the next long value.
	 *
	 * @return The next long value.
	 * @since 2018/11/02
	 */
	public long nextLong()
	{
		return ((long)this.next(32) << 32) + this.next(32);
	}
	
	/**
	 * Sets the seed for this random number generator.
	 *
	 * This is an atomic operation.
	 *
	 * @param __seed The seed to use.
	 * @since 2018/11/02
	 */
	public void setSeed(long __seed)
	{
		synchronized (this)
		{
			this._seed = (__seed ^ 0x5DEECE66DL) & ((1L << 48) - 1);
			this._hasnng = false;
		}
	}
	
	/**
	 * Strict logarithm implementation.
	 *
	 * @param __v The input value.
	 * @return The resulting logarithm value.
	 * @since 2018/11/02
	 */
	private static double __strictLog(double __v)
	{
		throw new todo.TODO();
	}
	
	/**
	 * Strict square root implementation, needed by {@link #nextGaussian()}.
	 *
	 * @param __v The input value.
	 * @return The resulting square root value.
	 * @since 2018/11/02
	 */
	private static double __strictSqrt(double __v)
	{
		throw new todo.TODO();
	}
}

File: ratufacoat/src/random.c
/* -*- Mode: C; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// Multi-Phasic Applications: SquirrelJME
//     Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the GNU General Public License v3+, or later.
// See license.mkd for licensing and copyright information.
// -------------------------------------------------------------------------*/

#include "debug.h"
#include "random.h"

sjme_jboolean sjme_randomSeed(sjme_randomState* random, sjme_jlong seed,
	sjme_error* error)
{
	sjme_todo("Fix this");
	return sjme_false;
#if 0
	/* Check. */
	if (random == NULL)
	{
		sjme_setError(error, SJME_ERROR_NULLARGS, 0);
		return sjme_false;
	}
	
	/* Wipe the random state. */
	memset(random, 0, sizeof(*random));
	
	/* Calculate the resultant seed. */
	/* this._seed = (__seed ^ 0x5_DEECE66DL) & ((1L << 48) - 1); */
	random->seed.hi = (seed.hi ^ SJME_JINT_C(0x00000005)) & 0xFFFF;
	random->seed.lo = (seed.lo ^ SJME_JINT_C(0xDEECE66D));
		
	return sjme_true;
#endif
}

sjme_jboolean sjme_randomNextBoolean(sjme_randomState* random,
	sjme_jboolean* out, sjme_error* error)
{
	sjme_jint val;
	
	if (out == NULL)
	{
		sjme_setError(error, SJME_ERROR_NULLARGS, 0);
		return sjme_false;
	}
	
	/* Get next value. */
	if (sjme_randomNextBits(random, &val, 1, error))
		return sjme_keepErrorF(error, SJME_ERROR_COULD_NOT_SEED, 0);
	
	*out = (val != 0 ? sjme_true : sjme_false);
	return sjme_true;
}

sjme_jboolean sjme_randomNextInt(sjme_randomState* random, sjme_jint* out,
	sjme_error* error)
{
	sjme_jint val;
	
	if (out == NULL)
	{
		sjme_setError(error, SJME_ERROR_NULLARGS, 0);
		return sjme_false;
	}
	
	/* Get next value. */
	if (sjme_randomNextBits(random, &val, 32, error))
		return sjme_keepErrorF(error, SJME_ERROR_COULD_NOT_SEED, 0);
	
	*out = val;
	return sjme_true;
}

sjme_jboolean sjme_randomNextBits(sjme_randomState* random, sjme_jint* out,
	sjme_jint bits, sjme_error* error)
{
	sjme_todo("Fix this");
	return sjme_false;
#if 0
	sjme_jlong seed;
	sjme_jint sh;
	
	/* Cannot be null. */
	if (random == NULL || out == NULL)
	{
		sjme_setError(error, SJME_ERROR_NULLARGS, 0);
		return sjme_false;
	}
	
	/* Must be in 32-bit range. */
	if (bits < 0 || bits > 32)
	{
		sjme_setError(error, SJME_ERROR_INVALID_ARGUMENT, bits);
		return sjme_false;
	}
	
	/* long seed = (this._seed * 0x5DEECE66DL + 0xBL) & ((1L << 48) - 1); */
	/* aa = this._seed * 0x5DEECE66DL. */
	seed = random->seed;
	seed = sjme_mulLongF(seed,
		SJME_JINT_C(0xDEECE66D), SJME_JINT_C(0x5));
	/* bb = aa + 0xBL */
	seed = sjme_addLongF(seed, SJME_JINT_C(0xB), 0);
	/* cc = & ((1L << 48) - 1) */
	seed.hi = seed.hi & SJME_JINT_C(0xFFFF);
	
	/*this._seed = seed; */
	random->seed.hi = seed.hi;
	random->seed.lo = seed.lo;
	
	/* return (int)(seed >>> (48 - __bits)); */
	sh = 48 - bits;
	if (sh >= 32)
		*out = seed.hi >> (sh - 32);
	else
		*out = (seed.hi << (32 - sh)) | sjme_ushrInt(seed.lo, sh);
	
	return sjme_true;
#endif
}
